sdks/java/harness/src/test/java/org/apache/beam/fn/harness/debug/DataSamplerTest.java

```diff
@@ -193,4 +221,44 @@ public void testFiltersMultiplePCollectionIds() throws Exception {
     assertHasSamples(samples, "a", ImmutableList.of(encodeString("a1"), encodeString("a2")));
     assertHasSamples(samples, "c", ImmutableList.of(encodeString("c1"), encodeString("c2")));
   }
+
+  /**
+   * Test that samples can be taken from the DataSampler while adding new OutputSamplers. This fails
+   * with a ConcurrentModificationException if there is a bug.
+   *
+   * @throws Exception
+   */
+  @Test
+  public void testConcurrentNewSampler() throws Exception {
+    DataSampler sampler = new DataSampler();
+    VarIntCoder coder = VarIntCoder.of();
+
+    // Create a thread that constantly creates new samplers.
+    Thread sampleThread =
+        new Thread(
+            () -> {
+              for (int i = 0; i < 1000000; i++) {
+                sampler.sampleOutput("pcollection-" + i, coder).sample(0);
+
+                // This sleep is here to allow for the test to stop this thread.
+                try {
+                  Thread.sleep(0);
+                } catch (InterruptedException e) {
+                  return;
+                }
+              }
+            });
+
+    sampleThread.start();
+
+    for (int i = 0; i < 20; i++) {
+      sampler.handleDataSampleRequest(
```

Comment:
This tests concurrency between a single sampling thread and getting the progress request.

You'll want to update this test to cover multiple threads (e.g. like 100) all creating the same set of 100 output samplers. You can have them all wait on a CountDownLatch which you release from the test thread before the output sampler creation starts allowing for all the threads to be ready to go (as done in https://github.com/apache/beam/blob/679d30256c6bd64d9760702c667d7d355e70166b/sdks/java/harness/src/test/java/org/apache/beam/fn/harness/control/ExecutionStateSamplerTest.java#L94).

At the same time you should continuously call handleDataSampleRequest until all the threads state has transitioned to TERMINATED (note that checking isAlive() can expose you to a race since start() doesn't mean that the thread is alive, just that the thread is scheduled to become alive at some point in the future). You can also ensure it is alive if you use another CountDownLatch to block the test thread from advancing to check if the sampler creating threads are alive as well. There are also futures as well.

Reply:
Cool, used a CountDownLatch